### PR TITLE
41 api to get transactions

### PR DESCRIPTION
**Pull request by rscronce**

### CODE REVIEW

**Review by sarabjotanand — APPROVED** [review]

File: backend/src/handler.py [hunk 1]
```diff
@@ -61,13 +61,13 @@ def _get_instances_from_rules(event, context):
...
-        'set_aside': '',
-        'biweekly_start': '',
-    })
+        parameters.update({
+            'start': '',
+            'end': '',
+            'current': '',
+            'set_aside': '',
+            'biweekly_start': '',
+        })
```
Comment (jamesfulford) [inline backend/src/handler.py:70]: ??? Isn't this fine?

File: backend/src/tests.py [hunk 2]
```diff
@@ -148,3 +150,26 @@ def test_rrule_validation_failure(self):
...
+        self.assertEqual(response.status_code, 400)
+
+    def test_future_limit_start_date(self):
+        now = datetime.now().date()
+        startDate = now + relativedelta(years=3, days=1)
+        endDate = now + relativedelta(years=3, days=2)        
+        response = self.client.post("http://testserver/api/transactions", params={"userid": "testuser", "currentBalance": "0", "startDate": startDate.strftime("%Y-%m-%d"), "endDate": endDate.strftime("%Y-%m-%d")})
+        self.assertEqual(response.status_code, 400)
+
+    def test_future_limit_end_date(self):
```
Comment (jamesfulford) [inline backend/src/tests.py:171]: Need to test missing query parameters and no query parameters.

I believe the validations you are doing are done in handler.py. Could we refactor to take out the stuff in handler.py? Calling handler was a hack on my part.

File: backend/src/views.py [hunk 3]
```diff
@@ -95,38 +97,56 @@ def delete_rule(request, rule_id, userid):
...
     return Response(None, status=status.HTTP_204_NO_CONTENT)
 
+def get_transaction_formatted_rule_list(userid):
+    rules = Rule.objects.filter(userid=userid)
+    rrules = RuleSerializer(rules, many=True).data
+    jsonBodyForTransactionQuery = {}
+
+    for rrule in rrules:
+        #TODO Check for duplicate keys and reconcile.
+        jsonBodyForTransactionQuery[rrule['name']] = {
```
Comment (jamesfulford) [inline backend/src/views.py:107]: Put a TODO to switch to 'id' instead of 'name' when the UI is ready for it.

File: backend/src/views.py [hunk 4]
```diff
@@ -95,38 +97,56 @@ def delete_rule(request, rule_id, userid):
...
+            "value": float(rrule['value'])
+        }
+
+    return json.dumps(jsonBodyForTransactionQuery, cls=DjangoJSONEncoder)
+
 @api_view(['GET'])
 def hello_world(request):
     return Response({ "status": "UP" })
 
 @api_view(['POST'])
```
Comment (jamesfulford) [inline backend/src/views.py:118]: Needs to be "GET"

File: backend/src/views.py [hunk 5]
```diff
@@ -95,38 +97,56 @@ def delete_rule(request, rule_id, userid):
...
 def process_transactions(request):
-    start = dt(2020, 10, 29)
-    end = dt(2021, 10, 29)
-    biweekly_start = dt(2020, 10, 29)
+    userId = request.GET.get('userid', '')
+    currentBalanceParam = request.GET.get('currentBalance', '')
```
Comment (jamesfulford) [inline backend/src/views.py:121]: Default is 0.

File: docker-compose.override.dev.yml [hunk 6]
```diff
@@ -5,6 +5,13 @@ networks:
     name: moneywise
 
 services:
+  moneywise-ui:
+    environment: 
+      - API_BASE_URL=http://localhost:8000
+    build:
+      args:
+        - API_BASE_URL=http://localhost:8000
```
Comment (jamesfulford) [inline docker-compose.override.dev.yml:13]: `API_BASE_URL` ain't a thing. Just use `REACT_APP_MONEYWISE_BASE_URL`.

File: docker-compose.override.dev.yml [hunk 7]
```diff
@@ -5,6 +5,13 @@ networks:
     name: moneywise
 
 services:
+  moneywise-ui:
+    environment: 
+      - API_BASE_URL=http://localhost:8000
```
Comment (jamesfulford) [inline docker-compose.override.dev.yml:10]: nginx doesn't need API_BASE_URL environment variable (or any, for that matter). Take out `environment` from this, please

File: backend/src/tests.py (hunk 2 above)
Comment (jamesfulford) [reply backend/src/tests.py:171]: (The handler refactor is not a required ask, but query params is required for the story. You decide whether it's this PR or not.)

File: backend/src/tests.py [hunk 8]
```diff
@@ -148,3 +150,26 @@ def test_rrule_validation_failure(self):
...
+        self.assertEqual(response.status_code, 400)     
+
+    def test_start_date_preceeds_end_date(self):
+        now = datetime.now().date()
+        endDate = now - relativedelta(days=1)      
+        response = self.client.post("http://testserver/api/transactions", params={"userid": "testuser", "currentBalance": "0", "startDate": now.strftime("%Y-%m-%d"), "endDate": endDate.strftime("%Y-%m-%d")})
+        self.assertEqual(response.status_code, 400)
+
+    def test_future_limit_start_date(self):
+        now = datetime.now().date()
```
Comment (jamesfulford) [inline backend/src/tests.py:165]: You're probably looking for `date.today()`

File: backend/src/handler.py (hunk 1 above)
Comment (rscronce) [reply backend/src/handler.py:70]: When the "parameters.update" code is not indented under the "if parameters is none:" statement, it has the effect of setting parameter values provided by the user to ''.

File: backend/src/tests.py (hunk 8 above)
Comment (rscronce) [reply backend/src/tests.py:165]: I've modified the code to use "date.today()"

File: backend/src/views.py (hunk 5 above)
Comment (rscronce) [reply backend/src/views.py:121]: Defaulted to '0' instead of ''.

File: backend/src/views.py (hunk 4 above)
Comment (rscronce) [reply backend/src/views.py:118]: Changed to @api_view(['GET'])

File: backend/src/views.py (hunk 3 above)
Comment (rscronce) [reply backend/src/views.py:107]: Done

File: docker-compose.override.dev.yml (hunk 7 above)
Comment (rscronce) [reply docker-compose.override.dev.yml:10]: Done

File: docker-compose.override.dev.yml (hunk 6 above)
Comment (rscronce) [reply docker-compose.override.dev.yml:13]: Done

File: backend/src/tests.py (hunk 2 above)
Comment (rscronce) [reply backend/src/tests.py:171]: Testing missing and no query parameters has been addressed in the latest push.

**Review by sarabjotanand — APPROVED** [review]

File: backend/src/tests.py (hunk 8 above)
Comment (jamesfulford) [reply backend/src/tests.py:193]: @rscronce did you forget to push? I don't see the change.

File: backend/src/views.py [hunk 9]
```diff
@@ -95,38 +97,75 @@ def delete_rule(request, rule_id, userid):
     rule.delete()
     return Response(None, status=status.HTTP_204_NO_CONTENT)
 
+def get_transaction_formatted_rule_list(userid):
+    rules = Rule.objects.filter(userid=userid)
+    rrules = RuleSerializer(rules, many=True).data
+    jsonBodyForTransactionQuery = {}
+
+    for rrule in rrules:
+        #TODO Check for duplicate keys and reconcile.
```
Comment (jamesfulford) [inline backend/src/views.py:106]: Can remove this comment.

File: backend/src/views.py [hunk 10]
```diff
@@ -95,38 +97,75 @@ def delete_rule(request, rule_id, userid):
...
 
     results = get_instances_from_rules({
         "queryStringParameters": {
             "start": start.strftime("%Y-%m-%d"),
             "end": end.strftime("%Y-%m-%d"),
-            "current": "1000",
+            "current": currentBalanceParam,
             "set_aside": "0",
-            "biweekly_start": biweekly_start.strftime("%Y-%m-%d")
+            "biweekly_start": start.strftime("%Y-%m-%d")
```
Comment (jamesfulford) [inline backend/src/views.py:166]: I'm not convinced this is needed anymore. Does it do anything in the code?

File: docker-compose.override.dev.yml [hunk 11]
```diff
@@ -5,6 +5,12 @@ networks:
     name: moneywise
 
 services:
+  moneywise-ui:
+    environment: 
+      - REACT_APP_MONEYWISE_BASE_URL=http://localhost:8000
```
Comment (jamesfulford) [inline docker-compose.override.dev.yml:10]: Again, `environment` is not needed for moneywise UI. It's only needed at build time, which is what build args is doing. Can remove this section.

File: docker-compose.override.dev.yml (hunk 7 above)
Comment (jamesfulford) [reply docker-compose.override.dev.yml]: Not quite, see new comment.

File: backend/src/tests.py (hunk 8 above)
Comment (jamesfulford) [reply backend/src/tests.py:193]: (And in the API code itself)

File: backend/src/views.py (hunk 9 above)
Comment (jamesfulford) [reply backend/src/views.py]: Nah, I got your back: 45fd21c52f10b23f4ee4f061eb710d51d0480688

File: docker-compose.override.dev.yml (hunk 11 above)
Comment (jamesfulford) [reply docker-compose.override.dev.yml:10]: Actually, even better: removing this. Having a UI container locally has not been helpful, has actually hurt my ability to start coding.

Comment (jamesfulford) [reply docker-compose.override.dev.yml]: Done: 8d0a29ee6760924964d7bf92acaa4b3ac6079303

File: backend/src/tests.py (hunk 8 above)
Comment (jamesfulford) [reply backend/src/tests.py]: Done: 4863f150d5dfd245752e298fce3f2ed69da6b015

**Review by jamesfulford — APPROVED** [review]
Resolved the remaining issues. Still not sure about biweekly_start, but I suppose we'll let it be

**Pull request merged**